lib/ansible/modules/windows/win_rds_cap.ps1

```diff
@@ -0,0 +1,323 @@
+$user_groups = Get-AnsibleParam -obj $params -name "user_groups" -type "list"
+$computer_groups = Get-AnsibleParam -obj $params -name "computer_groups" -type "list"
+
+# Device redirections
+$redirect_clipboard = Get-AnsibleParam -obj $params -name "redirect_clipboard" -type "bool"
+$redirect_drives = Get-AnsibleParam -obj $params -name "redirect_drives" -type "bool"
+$redirect_printers = Get-AnsibleParam -obj $params -name "redirect_printers" -type "bool"
+$redirect_serial = Get-AnsibleParam -obj $params -name "redirect_serial" -type "bool"
+$redirect_pnp = Get-AnsibleParam -obj $params -name "redirect_pnp" -type "bool"
+
+
+function Get-CAP([string] $name) {
+    $cap_path = "RDS:\GatewayServer\CAP\$name"
+    $cap = @{
+        Name = $name
+    }
+
+    # Fetch CAP properties
+    Get-ChildItem -Path "$cap_path" | foreach { $cap.Add($_.Name,$_.CurrentValue) }
+    # Convert boolean values
+    $cap.Enabled = $cap.Status -eq 1
+    $cap.Remove("Status")
+    $cap.AllowOnlySDRTSServers = $cap.AllowOnlySDRTSServers -eq 1
+
+    # Convert multiple choices values
+    $cap.AuthMethod = $auth_methods_set[$cap.AuthMethod]
+    $cap.SessionTimeoutAction = $session_timeout_actions_set[$cap.SessionTimeoutAction]
+
+    # Fetch CAP device redirection settings
+    $cap.DeviceRedirection = @{}
+    Get-ChildItem -Path "$cap_path\DeviceRedirection" | foreach { $cap.DeviceRedirection.Add($_.Name, ($_.CurrentValue -eq 1)) }
+
+    # Fetch CAP user and computer groups
+    $cap.UserGroups = @(Get-ChildItem -Path "$cap_path\UserGroups" | Select -ExpandProperty Name)
+    $cap.ComputerGroups = @(Get-ChildItem -Path "$cap_path\ComputerGroups" | Select -ExpandProperty Name)
+
+    return $cap
+}
+
+function Set-CAPPropertyValue {
+    [CmdletBinding(SupportsShouldProcess=$true)]
+    param (
+        [Parameter(Mandatory=$true)]
+        [string] $name,
+        [Parameter(Mandatory=$true)]
+        [string] $property,
+        [Parameter(Mandatory=$true)]
+        $value,
+        [Parameter()]
+        $resultobj = @{}
+    )
+
+    $cap_path = "RDS:\GatewayServer\CAP\$name"
+
+    try {
+        Set-Item -Path "$cap_path\$property" -Value $value -ErrorAction Stop
+    } catch {
+        Fail-Json -obj $resultobj -message "Failed to set property $property of CAP ${name}: $($_.Exception.Message)"
+    }
+}
+
+$result = @{
+  changed = $false
+}
+$diff_text = $null
+
+# Validate CAP name
+if ($name -match "[*/\\;:?`"<>|\t]+") {
+    Fail-Json -obj $result -message "Invalid character in CAP name."
+}
+
+# Validate user groups
+if ($null -ne $user_groups) {
+    if ($user_groups.Count -lt 1) {
+        Fail-Json -obj $result -message "Parameter 'user_groups' cannot be an empty list."
+    }
+
+    $user_groups = $user_groups | foreach {
+        $group = $_
+        # Test that the group is resolvable on the local machine
+        $sid = Convert-ToSID -account_name $group
```

Comment:
This will actually call `Fail-Json` if it is unable to resolve the SID. Did you come across a situation where this didn't happen?

Reply:
No I didn't, it's just a copy/paste fail from other modules, like [win_owner#L25-L28](https://github.com/ansible/ansible/blob/277bb3066c7ae8506a8201f5285b77d51b422267/lib/ansible/modules/windows/win_owner.ps1#L25-L28) and [win_share#L26-L29](https://github.com/ansible/ansible/blob/277bb3066c7ae8506a8201f5285b77d51b422267/lib/ansible/modules/windows/win_share.ps1#L26-L29), where I took inspiration.
Also [the error message verified in the tests](https://github.com/ansible/ansible/blob/77753a2b18d9b7f96f393e0d0c0b0ad4687dbd61/test/integration/targets/win_rds_cap/tasks/tests.yml#L254) is the one from the cmdlet, and not the one in the module.

That said, I think it might be a little better to leave the `Fail-Json` call in the module side in this case, to let the calling code throw the error with a more specific message, depending on the context. Currently the message will be `account_name [...] is not a valid account`, while we are actually dealing with groups. The message `[...] is not a valid computer group [...]` is a little bit more precise here.
That's not a big deal, but maybe we should add a silent flag to `Convert-ToSID` to allow the calling code to handle errors (or even ignore it), or reword the embedded error message, or something else... ?

Reply:
Interesting it must be throwing the error earlier than what is being caught in the cmdlet. I agree we should move it out or just have the module until throw and exception instead for finer control in the module.

Reply:
I just saw that there is also one place where the call to Convert-ToSID is surrounded by a try/catch block:
https://github.com/ansible/ansible/blob/277bb3066c7ae8506a8201f5285b77d51b422267/lib/ansible/modules/windows/win_audit_rule.ps1#L60-L66
I don't think there is an unhandled error in the cmdlet that would need to be caught in the calling code.

I suggest that we simply replace the call to `Fail-Json` in `Convert-ToSID` by `return $null`, to let the calling code test for `$null` and throw an error if needed:
```diff
@@ -79,13 +79,13 @@ Function Convert-ToSID {
     try {
         $account_sid = $account.Translate([System.Security.Principal.SecurityIdentifier])
     } catch {
-        Fail-Json @{} "account_name $account_name is not a valid account, cannot get SID: $($_.Exception.Message)"
+        return $null
     }

     return $account_sid.Value
 }
```
Do you agree with that?

Reply:
I would investigate this in another issue/pr. My original comment was merely an observation and don’t wish to make a change to a module_util that could affect other modules.